engine/crates/engine-v2/schema/src/builder/ids.rs

```diff
@@ -0,0 +1,186 @@
+//! This module is responsible for mapping identifiers between federated schema and the engine v2
+//! schema when the mapping is not 1:1.
+//!
+//! As of 3b56f12c95d334ce6cb46f4f8654ce531a69f975, this only happens when @inaccessible items
+//! are removed.
+
+use crate::ids::*;
+
+pub(super) trait MappableId: Copy {
+    type Counterpart: Copy + From<usize>;
+
+    fn get_idx(self) -> usize;
+    fn get_map(mapper: &IdMapper) -> &[usize];
+    fn get_map_mut(mapper: &mut IdMapper) -> &mut Vec<usize>;
+}
+
+macro_rules! mapped_ids {
+    ($($name:ident : $from:ty => $to:ty)*) => {
+
+        #[derive(Default, Debug)]
+        pub(super) struct IdMapper {
+            $(
+                /// The index of skipped ids.
+                $name: Vec<usize>,
+            )*
+        }
+
+        $(
+            impl MappableId for $from {
+                type Counterpart = $to;
+
+                fn get_idx(self) -> usize {
+                    self.0
+                }
+
+                fn get_map(mapper: &IdMapper) -> &[usize] {
+                    &mapper.$name
+                }
+
+                fn get_map_mut(mapper: &mut IdMapper) -> &mut Vec<usize> {
+                    &mut mapper.$name
+                }
+            }
+        )*
+    };
+}
+
+mapped_ids!(
+    fields: federated_graph::FieldId => FieldId
+    input_values: federated_graph::InputValueDefinitionId => InputValueDefinitionId
+    enum_values: federated_graph::EnumValueId => EnumValueId
+);
+
+impl IdMapper {
+    pub(super) fn skip<Id: MappableId>(&mut self, id: Id) {
+        let idx = id.get_idx();
+        let entries = Id::get_map_mut(self);
+
+        if let Some(last_entry) = entries.last().copied() {
+            assert!(last_entry < idx);
+        }
+
+        entries.push(idx);
+    }
+
+    pub(super) fn map<Id: MappableId>(&self, id: Id) -> Option<Id::Counterpart> {
+        let idx = id.get_idx();
+        let map = Id::get_map(self);
+        let skipped = map.partition_point(|skipped| *skipped <= idx);
+
+        if let Some(last) = map[..skipped].last().copied() {
+            if last == idx {
+                return None;
+            }
+        }
+
+        Some(Id::Counterpart::from(idx - skipped))
+    }
+
+    pub(super) fn map_range<Id: MappableId>(&self, (start_id, len): (Id, usize)) -> crate::IdRange<Id::Counterpart>
+    where
+        usize: From<Id::Counterpart>,
+    {
+        let start_idx = start_id.get_idx();
+        let skipped = Id::get_map(self);
+
+        // How many ids were skipped before the range.
+        let skipped_before = skipped.partition_point(|skipped| *skipped < start_idx);
+
+        // How many ids were skipped inside the range.
+        let skipped_inside = skipped[skipped_before..]
+            .iter()
+            .take_while(|skipped| **skipped < (start_idx + len))
+            .count();
```

Comment:
We can do a binary search here too, theoretically the complexity would be lower but I'd bet iterating and counting is going to be a lot faster in all realistic cases.